sdks/go/pkg/beam/core/runtime/harness/harness.go

```diff
@@ -401,21 +402,47 @@ func (c *control) handleInstruction(ctx context.Context, req *fnpb.InstructionRe
 				c.plans[bdID] = append(c.plans[bdID], plan)
 			}
 		}
+
+		// Check if the underlying DoFn self-checkpointed.
+		sr, delay, checkpointed, checkErr := plan.Checkpoint()
+
+		var rRoots []*fnpb.DelayedBundleApplication
+		if checkpointed {
+			rRoots = make([]*fnpb.DelayedBundleApplication, len(sr.RS))
+			for i, r := range sr.RS {
+				rRoots[i] = &fnpb.DelayedBundleApplication{
+					Application: &fnpb.BundleApplication{
+						TransformId:      sr.TId,
+						InputId:          sr.InId,
+						Element:          r,
+						OutputWatermarks: sr.OW,
+					},
+					RequestedTimeDelay: durationpb.New(delay),
+				}
+			}
+		}
+
 		delete(c.active, instID)
 		if removed, ok := c.inactive.Insert(instID); ok {
 			delete(c.failed, removed) // Also GC old failed bundles.
 		}
 		delete(c.metStore, instID)
+
 		c.mu.Unlock()
 
 		if err != nil {
 			return fail(ctx, instID, "process bundle failed for instruction %v using plan %v : %v", instID, bdID, err)
 		}
 
+		if checkErr != nil {
+			return fail(ctx, instID, "process bundle failed for instruction %v using plan %v : %v", instID, bdID, checkErr)
```

Comment:
Consider making it explicit that this error happened while checkpointing.

Reply:
Done